tools/libSwiftScan/SwiftCaching.cpp

```diff
@@ -0,0 +1,923 @@
+  };
+  auto notfound = [&]() {
+    return callback(ctx, nullptr, swift::c_string_utils::create_null());
+  };
+
+  auto ID = CAS.parseID(key);
+  if (!ID)
+    return failure(ID.takeError());
+
+  Cache.getAsync(*ID, globally,
+                 [&](llvm::Expected<std::optional<llvm::cas::CASID>> Result) {
+                   if (!Result)
+                     return failure(Result.takeError());
+                   if (!*Result)
+                     return notfound();
+
+                   auto Cached = createCachedCompilation(CAS, **Result, *ID);
+                   if (!Cached)
+                     return failure(Cached.takeError());
+
+                   if (!*Cached)
+                     return notfound();
+
+                   return callback(ctx, wrap(*Cached),
+                                   swift::c_string_utils::create_null());
+                 });
+}
+
+unsigned swiftscan_cached_compilation_get_num_outputs(
+    swiftscan_cached_compilation_t id) {
+  auto *Comp = unwrap(id);
+  if (auto *Result =
+          std::get_if<swift::cas::CompileJobCacheResult>(&Comp->Result))
+    return Result->getNumOutputs();
+
+  auto *Result = std::get_if<clang::cas::CompileJobCacheResult>(&Comp->Result);
+  assert(Result && "unexpected variant");
+  return Result->getNumOutputs();
+}
+
+swiftscan_output_kind_t
+swiftscan_cached_compilation_get_output_kind(swiftscan_cached_compilation_t id,
+                                             unsigned idx) {
+  auto *Comp = unwrap(id);
+  if (auto *Result =
+          std::get_if<swift::cas::CompileJobCacheResult>(&Comp->Result))
+    return getScanOutputKindFromFileID(Result->getOutput(idx).Kind);
+
+  auto *Result = std::get_if<clang::cas::CompileJobCacheResult>(&Comp->Result);
+  assert(Result && "unexpected variant");
+  switch (Result->getOutput(idx).Kind) {
+  case clang::cas::CompileJobCacheResult::OutputKind::MainOutput:
+    return SWIFTSCAN_OUTPUT_TYPE_CLANG_MODULE;
+  case clang::cas::CompileJobCacheResult::OutputKind::Dependencies:
+    return SWIFTSCAN_OUTPUT_TYPE_DEPENDENCIES;
+  case clang::cas::CompileJobCacheResult::OutputKind::SerializedDiagnostics:
+    return SWIFTSCAN_OUTPUT_TYPE_CACHED_DIAGNOSTICS;
+  }
+}
+
+swiftscan_cached_output_t
+swiftscan_cached_compilation_get_output(swiftscan_cached_compilation_t id,
+                                        unsigned idx) {
+  auto *Comp = unwrap(id);
+  if (auto *Result =
+          std::get_if<swift::cas::CompileJobCacheResult>(&Comp->Result))
+    return wrap(new SwiftCachedOutputHandle(Result->getOutput(idx).Object));
+
+  auto *Result = std::get_if<clang::cas::CompileJobCacheResult>(&Comp->Result);
+  assert(Result && "unexpected variant");
+  return wrap(new SwiftCachedOutputHandle(Result->getOutput(idx).Object));
+}
+
+void swiftscan_cached_compilation_dispose(swiftscan_cached_compilation_t id) {
+  delete unwrap(id);
+}
+
+bool swiftscan_cached_output_load(swiftscan_cas_t cas,
```

Comment:
To be sure that all the remote network work will be done in these functions and the node will be fully materialized once the functions complete, both `swiftscan_cached_output_load` and `swiftscan_cached_output_load_async` should walk the object node and make sure any potential sub-nodes are also loaded.
Similar to what `clang_experimental_cas_loadObjectByString` and `clang_experimental_cas_loadObjectByString_async` are doing.

I'd suggest to move the code from  the `clang_experimental_cas_loadObject*` functions to an `ObjectStore` API, something like `loadNodeTree/loadNodeTreeAsync`, and use it here as well.

Reply:
hmm, I did check the clang function. It is only loading the direct children, which might not be enough. Should this be a full tree traversal?

Reply:
The clang functions *are* doing a full tree traversal, did I do it wrong? Could you clarify why they are only doing direct children?

Reply:
My mistake. I probably read it wrong. But a full traversal is done in Swift API as well now.